go/vt/vttls/vttls.go

```diff
@@ -91,27 +97,97 @@ func ClientConfig(cert, key, ca, name string) (*tls.Config, error) {
 func ServerConfig(cert, key, ca string) (*tls.Config, error) {
 	config := newTLSConfig()
 
-	// Load the server cert and key.
-	crt, err := tls.LoadX509KeyPair(cert, key)
+	certificates, err := loadTLSCertificate(cert, key)
+
 	if err != nil {
-		return nil, fmt.Errorf("failed to load cert/key: %v", err)
+		return nil, err
 	}
-	config.Certificates = []tls.Certificate{crt}
+
+	config.Certificates = *certificates
 
 	// if specified, load ca to validate client,
 	// and enforce clients present valid certs.
 	if ca != "" {
-		b, err := ioutil.ReadFile(ca)
+		certificatePool, err := loadx509CertPool(ca)
+
 		if err != nil {
-			return nil, fmt.Errorf("failed to read ca file: %v", err)
-		}
-		cp := x509.NewCertPool()
-		if !cp.AppendCertsFromPEM(b) {
-			return nil, fmt.Errorf("failed to append certificates")
+			return nil, err
 		}
-		config.ClientCAs = cp
+
+		config.ClientCAs = certificatePool
 		config.ClientAuth = tls.RequireAndVerifyClientCert
 	}
 
 	return config, nil
 }
+
+var certPools = make(map[string]*x509.CertPool)
```

Comment:
This really ought not be a map. While it is low probability there are a few places this is called in go-routines. You should convert it to a sync.Map or create a lock for Reading and Writing.

Reply:
I've push the change up